### PR TITLE
add xalgo consensus

### PR DESCRIPTION
**Pull request by gidonkatten**

### CODE REVIEW

File: src/xAlgo/allocationStrategies/greedy.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import { maximum, minimum } from "../../mathLib";
+import { ProposerAllocation, ConsensusState } from "../types";
+import { convertAlgoToXAlgoWhenDelay } from "../formulae";
+import { FIXED_CAPACITY_BUFFER, MAX_APPL_CALLS } from "./constants";
+
+const greedyStakeAllocationStrategy = (consensusState: ConsensusState, amount: number | bigint): ProposerAllocation => {
+  const { proposersBalances, maxProposerBalance } = consensusState;
+  const strategy = new Array<bigint>(proposersBalances.length);
```
Comment (stefanofa) [inline src/xAlgo/allocationStrategies/greedy.ts:8]:
```suggestion
  const allocations = new Array<bigint>(proposersBalances.length);
```

File: src/xAlgo/allocationStrategies/greedy.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  return strategy;
+};
+
+const greedyUnstakeAllocationStrategy = (
+  consensusState: ConsensusState,
+  amount: number | bigint,
+): ProposerAllocation => {
+  const { proposersBalances, minProposerBalance } = consensusState;
+  const strategy = new Array<bigint>(proposersBalances.length);
```
Comment (stefanofa) [inline src/xAlgo/allocationStrategies/greedy.ts:40]:
```suggestion
  const allocations = new Array<bigint>(proposersBalances.length);
```

File: src/xAlgo/types.ts [hunk 3]
```diff
@@ -1,17 +1,30 @@
...
+  totalPendingStake: bigint;
+  totalActiveStake: bigint;
+  totalRewards: bigint;
+  totalUnclaimedFees: bigint;
+  canImmediateStake: boolean;
+  canDelayStake: boolean;
 }
 
-export { XAlgo, XAlgoInfo };
+type ProposerAllocation = bigint[];
```
Comment (stefanofa) [inline src/xAlgo/types.ts:28]: Maybe it is better to call it:
`ProposerAllocations` in plural?

File: src/xAlgo/consensus.ts [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+ * @returns Transaction[] stake transactions
+ */
+function prepareImmediateStakeTransactions(
+  consensusConfig: ConsensusConfig,
+  consensusState: ConsensusState,
+  senderAddr: string,
+  amount: number | bigint,
+  minReceivedAmount: number | bigint,
+  params: SuggestedParams,
+  allocationStrategy = defaultStakeAllocationStrategy(consensusState, amount),
```
Comment (stefanofa) [inline src/xAlgo/consensus.ts:204]:
```suggestion
  proposerAllocations = defaultStakeAllocationStrategy(consensusState, amount),
```

File: src/xAlgo/consensus.ts [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+ * @param note - optional note to distinguish who is the minter (must pass to be eligible for revenue share)
+ * @returns Transaction[] stake transactions
+ */
+function prepareDelayedStakeTransactions(
+  consensusConfig: ConsensusConfig,
+  consensusState: ConsensusState,
+  senderAddr: string,
+  amount: number | bigint,
+  params: SuggestedParams,
+  allocationStrategy = defaultStakeAllocationStrategy(consensusState, amount),
```
Comment (stefanofa) [inline src/xAlgo/consensus.ts:260]:
```suggestion
  proposerAllocations = defaultStakeAllocationStrategy(consensusState, amount),
```

File: src/xAlgo/consensus.ts [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+ * @returns Transaction[] unstake transactions
+ */
+function prepareUnstakeTransactions(
+  consensusConfig: ConsensusConfig,
+  consensusState: ConsensusState,
+  senderAddr: string,
+  amount: number | bigint,
+  minReceivedAmount: number | bigint,
+  params: SuggestedParams,
+  allocationStrategy = defaultUnstakeAllocationStrategy(consensusState, amount),
```
Comment (stefanofa) [inline src/xAlgo/consensus.ts:368]:
```suggestion
  proposerAllocations = defaultUnstakeAllocationStrategy(consensusState, amount),
```

**Review by stefanofa — CHANGES_REQUESTED** [review]

File: src/xAlgo/allocationStrategies/greedy.ts (hunk 1 above)
Comment (gidonkatten) [reply src/xAlgo/allocationStrategies/greedy.ts]: cf950281a5de8984785b4be0be8b68d2ec854f6d

File: src/xAlgo/allocationStrategies/greedy.ts (hunk 2 above)
Comment (gidonkatten) [reply src/xAlgo/allocationStrategies/greedy.ts]: cf950281a5de8984785b4be0be8b68d2ec854f6d

File: src/xAlgo/consensus.ts (hunk 6 above)
Comment (gidonkatten) [reply src/xAlgo/consensus.ts]: 549c611c0cb7e101c895cb8f15e274d4b41b5f93

File: src/xAlgo/consensus.ts (hunk 5 above)
Comment (gidonkatten) [reply src/xAlgo/consensus.ts]: 549c611c0cb7e101c895cb8f15e274d4b41b5f93

File: src/xAlgo/consensus.ts (hunk 4 above)
Comment (gidonkatten) [reply src/xAlgo/consensus.ts]: 549c611c0cb7e101c895cb8f15e274d4b41b5f93

File: src/xAlgo/types.ts (hunk 3 above)
Comment (gidonkatten) [reply src/xAlgo/types.ts]: 549c611c0cb7e101c895cb8f15e274d4b41b5f93

**Review by palace22 — APPROVED** [review]

**Review by stefanofa — DISMISSED** [review]

**Pull request merged**